src/common/sv2_noise.cpp

```diff
@@ -0,0 +1,529 @@
+    size_t plain_position = 0;
+    std::vector<std::byte> ad; // No associated data
+
+    while (processed < ciphertext.size()) {
+        size_t chunk_size = std::min(ciphertext.size() - processed, NOISE_MAX_CHUNK_SIZE);
+        Span<std::byte> chunk_cipher = ciphertext.subspan(processed, chunk_size);
+        Span<std::byte> chunk_plain = plain.subspan(plain_position, chunk_size - Poly1305::TAGLEN);
+        if (!DecryptWithAd(ad, chunk_cipher, chunk_plain)) return false;
+        processed += chunk_size;
+        plain_position += chunk_size - Poly1305::TAGLEN;
+    }
+
+    return true;
+}
+
+void Sv2SymmetricState::MixHash(const Span<const std::byte> input)
+{
+    m_hash_output = (HashWriter{} << m_hash_output << input).GetSHA256();
+}
+
+void Sv2SymmetricState::MixKey(const Span<const std::byte> input_key_material)
+{
+    uint8_t out0[Sv2CipherState::HASHLEN], out1[Sv2CipherState::HASHLEN];
+
+    HKDF2(input_key_material, out0, out1);
+
+    std::memset(m_chaining_key, 0, sizeof(m_chaining_key));
+    std::copy(out0, out0 + Sv2CipherState::HASHLEN, m_chaining_key);
+    m_cipher_state = Sv2CipherState{out1};
+}
+
+std::string Sv2SymmetricState::GetChainingKey()
+{
+    return HexStr(m_chaining_key);
+}
+
+void Sv2SymmetricState::LogChainingKey()
+{
+    LogTrace(BCLog::SV2, "Chaining key: %s\n", GetChainingKey());
+}
+
+void Sv2SymmetricState::HKDF2(const Span<const std::byte> input_key_material, uint8_t out0[Sv2CipherState::HASHLEN], uint8_t out1[Sv2CipherState::HASHLEN])
+{
+    uint8_t tmp_key[Sv2CipherState::HASHLEN];
+    CHMAC_SHA256 tmp_mac(m_chaining_key, Sv2CipherState::HASHLEN);
+    tmp_mac.Write(UCharCast(input_key_material.data()), input_key_material.size());
+    tmp_mac.Finalize(tmp_key);
+
+    CHMAC_SHA256 out0_mac(tmp_key, Sv2CipherState::HASHLEN);
+    uint8_t one[1]{0x1};
+    out0_mac.Write(one, 1);
+    out0_mac.Finalize(out0);
+
+    std::vector<uint8_t> in1;
+    in1.reserve(Sv2CipherState::HASHLEN + 1);
+    std::copy(out0, out0 + Sv2CipherState::HASHLEN, std::back_inserter(in1));
+    in1.push_back(0x02);
+
+    CHMAC_SHA256 out1_mac(tmp_key, Sv2CipherState::HASHLEN);
+    out1_mac.Write(&in1[0], in1.size());
+    out1_mac.Finalize(out1);
+}
+
+bool Sv2SymmetricState::EncryptAndHash(Span<const std::byte> plain, Span<std::byte> ciphertext)
+{
+    Assume(Sv2Cipher::EncryptedMessageSize(plain.size()) == ciphertext.size());
+
+    if (!m_cipher_state.EncryptWithAd(MakeByteSpan(m_hash_output), plain, ciphertext)) {
+        return false;
+    }
+    MixHash(ciphertext);
+    return true;
+}
+
+bool Sv2SymmetricState::DecryptAndHash(Span<std::byte> ciphertext, Span<std::byte> plain)
+{
+    Assume(Sv2Cipher::EncryptedMessageSize(plain.size()) == ciphertext.size());
+
+    // The handshake requires mix hashing the cipher text NOT the decrypted
+    // plaintext.
+    std::vector<std::byte> ciphertext_copy;
+    ciphertext_copy.assign(ciphertext.begin(), ciphertext.end()); // (ciphertext.size(), std::byte(0));
```

Comment:
nit: Consider removing the commented out code. Same to the following line as well.